Quantum-Control-Applications/ESR/pi_pulse_calibration.py

```diff
@@ -55,36 +60,37 @@
             reset_phase("resonator")
             reset_frame("ensemble")
 
-            echo = declare_stream(adc_trace=True)
-
+            # Set pulse2_len to 2*pulse1_len so that pulse2 is pi if pulse1 is pi/2
             assign(pulse2_len, 2 * pulse1_len)
+            # Set pulse_delay to safe_delay - center_to_center - internal_delay (found with simulation)
             assign(
                 pulse_delay,
-                2000 // 4 - Cast.mul_int_by_fixed(pulse1_len, 0.5) - Cast.mul_int_by_fixed(pulse2_len, 0.5) - 8,
+                safe_delay - Cast.mul_int_by_fixed(pulse1_len + pulse2_len, 0.5) - 8,
             )
+            # Set pulse_delay to safe_delay - center_to_center - internal_delay (found with simulation)
             assign(
                 readout_delay,
-                2000 // 4 - Cast.mul_int_by_fixed(pulse2_len, 0.5) - Cast.mul_int_by_fixed(readout_len, 0.125) - 5,
+                safe_delay - Cast.mul_int_by_fixed(pulse2_len + readout_len // 4, 0.5) - 5,
             )
-
-            # we delay the switches because `duration` in digital pulses
-            # takes less cycles to compute than in analog ones
-            wait(11, "switch_1", "switch_2")
+            # Play first pulse along X
+            play("const", "ensemble", duration=pulse1_len)
+            # we delay the switches because `duration` for digital pulses is faster than for analog
+            # We use the simulator to make the adjustments and find `8`
+            wait(8, "switch_1", "switch_2")
             play("activate", "switch_1", duration=pulse1_len)
             play("activate", "switch_2", duration=pulse1_len)
-            play("const", "ensemble", duration=pulse1_len)
-
+            # Desired delay
             wait(pulse_delay, "ensemble", "switch_1", "switch_2")
-
+            # Play second pulse along Y
             frame_rotation_2pi(-0.5, "ensemble")
             play("activate", "switch_1", duration=pulse2_len)
             play("activate", "switch_2", duration=pulse2_len)
             play("const", "ensemble", duration=pulse2_len)
 
             align()  # global align
-
+            # Wait some time to avoid sending pulses in the measurement window
```

Comment:
it is not only to avoid sending pulse in the measurement window. The delay time ideally is to place the recording window centered at the echo pulse. You could have an echo of 200 ns that appear 1 us after `pi`and a recording window of 500 ns so you want to delay the measuremnt so as to have the echo centered in the recording window

Reply:
Yes, much clearer after our discussion thanks!